fn.go

```diff
@@ -0,0 +1,230 @@
+func (x *Fn) Int() *big.Int {
+	ret := new(big.Int)
+	x.GetInt(ret)
+	return ret
+}
+
+// GetInt sets the given big.Int to be equal to the field element.
+func (x *Fn) GetInt(dst *big.Int) {
+	var bs [32]byte
+	x.GetB32(bs[:])
+	dst.SetBytes(bs[:])
+}
+
+// SetB32 sets the field element to be equal to the given byte slice,
+// interepreted as big endian. The field element will be reduced modulo N. The
+// argument `overflow` will be non zero if the bytes represented a number
+// greater than or equal to N, and zero otherwise.
+//
+// Panics: If the byte slice has length less than 32, this function will panic.
+func (x *Fn) SetB32(bs []byte) bool {
+	// 64 bits in case the c representation of an int has 64 bits.
+	var overflow int64
+
+	C.secp256k1_scalar_set_b32(
+		&x.inner,
+		(*C.uchar)(&bs[0]),
+		(*C.int)(unsafe.Pointer(&overflow)),
+	)
+	return overflow != 0
+}
+
+// SetB32SecKey sets the receiver from the given byte slice, intepreted in big
+// endian form, and returns a bool indicating whether the bytes represent a
+// valid private key. The bytes don't represent a valid private key if either
+// they represent a number greater than or equal to N, or if they represent the
+// zero element.
+//
+// Panics: If the byte slice has length less than 32, this function will panic.
+func (x *Fn) SetB32SecKey(bs []byte) bool {
+	return int(C.secp256k1_scalar_set_b32_seckey(&x.inner, (*C.uchar)(C.CBytes(bs)))) == 1
+}
+
+// GetB32 stores the bytes of the field element into destination in big endian
+// form.
+func (x Fn) GetB32(dst []byte) {
+	C.secp256k1_scalar_get_b32((*C.uchar)(&dst[0]), &x.inner)
+}
+
+// SizeHint implements the surge.SizeHinter interface.
+func (x Fn) SizeHint() int { return 32 }
+
+// Marshal implements the surge.Marshaler interface.
+func (x Fn) Marshal(w io.Writer, m int) (int, error) {
+	if m < 32 {
+		return m, surge.ErrMaxBytesExceeded
+	}
+
+	var bs [32]byte
+	x.GetB32(bs[:])
+	n, err := w.Write(bs[:])
+
+	return m - n, err
+}
+
+// Unmarshal implements the surge.Unmarshaler interface.
+func (x *Fn) Unmarshal(r io.Reader, m int) (int, error) {
+	if m < 32 {
+		return m, surge.ErrMaxBytesExceeded
+	}
+
+	var bs [32]byte
+	n, err := io.ReadFull(r, bs[:])
+	m -= n
+	if err != nil {
+		return m, err
+	}
+	x.SetB32(bs[:])
+
+	return m, nil
+}
+
+// SetUint sets the field element to be equal to the given uint.
+func (x *Fn) SetUint(v uint16) {
+	// TODO: Currently we take a uint16 as an argument because a c uint is only
+	// guaranteed to have at least 16 bits. Consider changing the struct to
+	// have x.inner be [4]uint64 to avoid this potential information loss.
+	C.secp256k1_scalar_set_int(&x.inner, C.uint(v))
+}
+
+// Add computes the addition of the two field elements and stores the result in
+// the receiver.
+func (x *Fn) Add(a, b *Fn) {
+	// The c function returns an int that indicates whether there was overflow,
+	// which we ignore.
+	_ = C.secp256k1_scalar_add(&x.inner, &a.inner, &b.inner)
+}
```

Comment:
What happens if `a` or `b` are `nil`? What happens if two or more of the pointers involved point to the same memory? Consider adding explicit `nil` checks and updating the comment.

Reply:
This same comment applies to all the arithmetic that follows (`Mul`, `Inverse`, etc.).

Reply:
Do these methods need `*Fn` or are they able to use `Fn`? If `&x.inner` is passed in by reference solely for efficiency, then an `Fn` receiver should be fine. If it is passed in by reference for some other reason, then `*Fn` is probably needed.